lightning/src/ln/channelmanager.rs

```diff
@@ -1604,15 +1622,49 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref> ChannelMan
 									htlc_id: prev_htlc_id,
 									incoming_packet_shared_secret: incoming_shared_secret,
 								};
-								channel_state.claimable_htlcs.entry(payment_hash).or_insert(Vec::new()).push(ClaimableHTLC {
+
+								let mut total_value = 0;
+								let payment_secret_opt =
+									if let &Some(ref data) = &payment_data { Some(data.payment_secret.clone()) } else { None };
+								let htlcs = channel_state.claimable_htlcs.entry((payment_hash, payment_secret_opt))
+									.or_insert(Vec::new());
```

Comment:
True, but sadly in general these types of issues are no more of a DoS concern than them just forwarding an HTLC through us and holding it open. In this specific case, though, I think we will send any such HTLC upstream as a PaymentReceived event. Certainly *not* doing so would be a privacy issue (as any handling of "same-hash-different-secret" HTLCs differently from "unique-hash" is inherently a privacy concern). Because we index on (hash, secret), not just hash, I dont see how, though.

Reply:
Right _not_ handling them uniquely would be a privacy leak, at least cancelling without any timeout. As you pointed out, it falls under the bigger issue of free pending HTLCs on channel links, should be revisited latter if we get upfront_payment or any other DoS prevention measure.